tests/virt_autotest/kubevirt_tests_server.pm

```diff
@@ -0,0 +1,255 @@
+    'console_test',
+    'credentials_test',
+    'expose',
+    'replicaset_test',
+    'network_services',
+    'vmi_kernel_boot_test',
+    'infra_test',
+    'operator_test',
+    'migration_test',
+    'subresource_api_test',
+    'virt_control_plane_test'
+);
+
+sub run {
+    my ($self) = @_;
+    my $ip_out = script_output('ip route show | grep -Eo "src\s+([0-9.]*)\s+" | head -1 | cut -d\' \' -f 2', 30);
+    set_var('SERVER_IP', $ip_out);
+    bmwqemu::save_vars();
+
+    # Wait for destination to be ready
+    $self->set_mutex_lock('DST_READY_TO_START');
+
+    my $agent_ip = $self->get_var_from_parent("AGENT_IP");
+
+    rke2_server_setup($self, $agent_ip);
+    deploy_kubevirt_manifests();
+    run_virt_tests();
+    generate_test_report();
+    upload_results();
+}
+
+# Mutex lock/unlock functionality
+sub set_mutex_lock {
+    my ($self, $lock_signal) = @_;
+
+    mutex_lock($lock_signal);
+    mutex_unlock($lock_signal);
+}
+
+sub rke2_server_setup {
+    my ($self, $agent_ip) = @_;
+
+    systemctl('stop apparmor.service');
+    # systemctl('stop SuSEfirewall2.service');
+    assert_script_run('ssh-keygen -b 2048 -t rsa -q -N "" -f ~/.ssh/id_rsa') unless (-e "~/.ssh/id_rsa");
+    exec_and_insert_password("ssh-copy-id -o StrictHostKeyChecking=no -i /root/.ssh/id_rsa.pub root\@$agent_ip");
+
+    # Ensure SUSE certificates are installed on the node
+    ensure_ca_certificates_suse_installed();
+
+    # Install downstream kubevirt packages
+    zypper_call('in kubernetes1.18-client') if (script_run('rpmquery kubernetes1.18-client'));
+
+    # RKE2 deployment on server node
+    my $rke_ver = get_var('INSTALL_RKE2_VERSION');
+    if ($rke_ver) {
+        assert_script_run("curl -sfL https://get.rke2.io | INSTALL_RKE2_VERSION=$rke_ver sh -");
+    } else {
+        assert_script_run('curl -sfL https://get.rke2.io | sh -');
+    }
+    systemctl('enable rke2-server.service');
+    systemctl('start rke2-server.service');
+
+    # Start rke2-server service ready
+    mutex_create('RKE2_SERVER_START_READY');
+
+    # Wait for rke2-agent service to be ready
+    $self->set_mutex_lock('RKE2_AGENT_START_READY');
+
+    assert_script_run('cp /etc/rancher/rke2/rke2.yaml ~/.kube/config');
+    assert_script_run('kubectl get nodes');
+
+    # Create registries ready
+    script_run("cat > /etc/rancher/rke2/registries.yaml <<__END
+mirrors:
+  registry.suse.de:
+    endpoint:
+      - http://registry.suse.de
+configs:
+  registry.suse.de:
+    tls:
+      insecure_skip_verify: true
+__END");
+
+    assert_script_run("scp /etc/rancher/rke2/registries.yaml root\@$agent_ip:/etc/rancher/rke2/registries.yaml");	
+    mutex_create('CREATE_REGISTRIES_READY');
+
+    # Restart RKE2 service and check the service is active well after restart
+    systemctl('restart rke2-server.service');
+    check_service_status();
+}
+
+sub check_service_status {
+    # Check RKE2 services & kubevirt test environment
+    systemctl('is-active rke2-server.service');
+}
+
```

Comment:
I think communication between agent and server can be improved a little bit:
1. Do not give the other end a signal to proceed if something goes wrong on itself.
2. Only give the other end a signal to proceed if everything goes well on itself.
3. Consider signal alarm to the other end if something goes wrong on itself.